File: Assignment/RootFindingMethods/bisection_method.cpp
#include <bits/stdc++.h>
using namespace std;
long double EPS;

long double func(long double x)
{
	return 2*x*x*x - 11*x - 50;
}

void bisection(long double a, long double b)
{
	if(0 <= func(a) * func(b))
	{
		cout << "The root doesn't exist between " << a << " and " << b << "." << endl;
		return;
	}
	long double c = a;
	while(b - a >= EPS)
	{
		c = (a + b) / 2;
		if(func(c) == 0)
			break;
		else if(func(a) * func(c) < 0)
			b = c;
		else
			a = c;
	}
	cout << "The root of f(x) = 2x^3 - 11x - 50 is : " << c << endl;
}

int main()
{
	long double a, b;
	cout << "Enter guessed a and b." << endl;
	cout << "a : ";
	cin >> a;
	cout << "b : ";
	cin >> b;
	cout << "Enter epsilon: ";
	cin >> EPS;
	bisection(a, b);
	return 0;
}

File: Assignment/RootFindingMethods/bisection_method.m
function bisection_method()
    f = @(x) x^3 - 2*x^2 - 4
    a = 2
    b = 3
    approximation = 1e-6
    eps = approximation;
    fprintf('\n\n')
    [solution, no_iterations] = Bisection(f, a, b, eps);
    if no_iterations > 0
		fprintf('Number of funtion calls: %d\n', no_iterations);
		fprintf('A solution is: %f\n', solution);
    else
		fprintf('Abort execution!\n');
	end
end

function [solution, no_iterations] = Bisection(f, a, b, eps)
    fprintf("%s\t%s\t\t\t%s\t\t\t%s\n", "Iteration", "a", "b", "root");
    no_iterations = 0;
    if f(a) * f(b) >= 0
        fprintf('Single root does not exist between %f and %f.\n', a, b);
        exit(1);
    endif
    solution = a;
    while abs(b - a) >= eps
        solution = (a + b) / 2;
        no_iterations = no_iterations + 1;
        if f(solution) == 0
            break;
        elseif (f(a) * f(solution)) < 0
            b = solution;
        else
            a = solution;
        end
        fprintf("%d\t\t%f\t\t%f\t\t%f\n", no_iterations, a, b, solution);
    end
end